scipy/linalg/decomp.py

```diff
@@ -371,126 +415,170 @@ def eigh(a, b=None, lower=True, eigvals_only=False, overwrite_a=False,
         b1 = _asarray_validated(b, check_finite=check_finite)
         overwrite_b = overwrite_b or _datacopied(b1, b)
         if len(b1.shape) != 2 or b1.shape[0] != b1.shape[1]:
-            raise ValueError('expected square matrix')
+            raise ValueError('expected square "b" matrix')
 
         if b1.shape != a1.shape:
-            raise ValueError("wrong b dimensions %s, should "
-                             "be %s" % (str(b1.shape), str(a1.shape)))
-        if iscomplexobj(b1):
-            cplx = True
-        else:
-            cplx = cplx or False
-    else:
-        b1 = None
+            raise ValueError("wrong b dimensions {}, should "
+                             "be {}".format(b1.shape, a1.shape))
 
-    # Set job for Fortran routines
-    _job = (eigvals_only and 'N') or 'V'
+        cplx = True if iscomplexobj(b1) else (cplx or False)
+        drv_args.update({'overwrite_b': overwrite_b})
 
-    # port eigenvalue range from Python to Fortran convention
-    if eigvals is not None:
-        lo, hi = eigvals
-        if lo < 0 or hi >= a1.shape[0]:
-            raise ValueError('The eigenvalue range specified is not valid.\n'
-                             'Valid range is [%s,%s]' % (0, a1.shape[0]-1))
-        lo += 1
-        hi += 1
-        eigvals = (lo, hi)
+    # backwards-compatibility handling
+    subset_by_index = subset_by_index if (eigvals is None) else eigvals
 
-    # set lower
-    if lower:
-        uplo = 'L'
-    else:
-        uplo = 'U'
+    subset = (subset_by_index is not None) or (subset_by_value is not None)
+
+    # Both subsets can't be given
+    if subset_by_index and subset_by_value:
+        raise ValueError('Either index or value subset can be requested.')
+
+    # Take turbo into account if all conditions are met otherwise ignore
+    if turbo:
+        if b is not None:
+            driver = 'gvx' if subset else 'gvd'
+
+    # Check indices if given
+    if subset_by_index:
+        lo, hi = [int(x) for x in subset_by_index]
+        if not (0 <= lo <= hi < n):
+            raise ValueError('Requested eigenvalue indices are not valid. '
+                             'Valid range is [0, {}] and start <= end, but '
+                             'start={}, end={} is given'.format(n-1, lo, hi))
+        # fortran is 1-indexed
+        drv_args.update({'range': 'I', 'il': lo + 1, 'iu': hi + 1})
+
+    if subset_by_value:
+        lo, hi = subset_by_value
+        if not (-inf <= lo < hi <= inf):
+            raise ValueError('Requested eigenvalue bounds are not valid. '
+                             'Valid range is (-inf, inf) and low < high, but '
+                             'low={}, high={} is given'.format(lo, hi))
+
+        drv_args.update({'range': 'V', 'vl': lo, 'vu': hi})
 
     # fix prefix for lapack routines
-    if cplx:
-        pfx = 'he'
+    pfx = 'he' if cplx else 'sy'
+
+    # decide on the driver if not given
+    # first early exit on incompatible choice
+    if driver:
```

Comment:
Should these be tested in `test_wrong_inputs`?

Reply:
You are right. I will add.